pkg/plugins/config-gen/v1alpha/scaffolds/internal/templates/makefile.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"sigs.k8s.io/kubebuilder/v3/pkg/machinery"
+)
+
+var _ machinery.Template = &Makefile{}
+
+// Makefile scaffolds a file that defines project management CLI commands
+type Makefile struct {
+	machinery.TemplateMixin
+	machinery.ComponentConfigMixin
+
+	// Image is controller manager image name
+	Image string
+	// BoilerplatePath is the path to the boilerplate file
+	BoilerplatePath string
+	// Controller tools version to use in the project
+	ControllerToolsVersion string
+	// Kustomize version to use in the project
+	KustomizeVersion string
+	// ControllerRuntimeVersion version to be used to download the envtest setup script
+	ControllerRuntimeVersion string
+}
+
+// SetTemplateDefaults implements file.Template
+func (f *Makefile) SetTemplateDefaults() error {
+	if f.Path == "" {
+		f.Path = "Makefile"
+	}
+
+	f.TemplateBody = makefileTemplate
+
+	if f.Image == "" {
+		f.Image = "controller:latest"
+	}
+
+	// FIXME: How can we customize makefile without overwrite it since it is language based?
+	f.IfExistsAction = machinery.OverwriteFile
+
+	return nil
+}
+
+//nolint:lll
+const makefileTemplate = `
```

Comment:
What should be diff for config-gen in the Makefile?
Have we anything that is specific for config-gen that we need to change?
Have we anything that we should remove ?

Reply:
c/c @pwittrock

Reply:
Hey @camilamacedo86 , thx for the review!

So this is the main part that confuses me of how such `alpha config-gen` get migrated to the plugin and worked by the integration with language specific plugins.

Because, once it becomes a plugin, the function of `config-gen` can't be actively triggered.
Especially when we'd like to use it as:
```
kubebuilder init --plugins=config-gen/v1-alpha,base.go.kubebuilder.io/v3 --domain example.org --repo example.org/guestbook-operator 
```

I understand that we don't expect the plugin to "update" makefile, as that is defined as "language specific".
And, it seems that most of the makefile templates(go/v3, go/v2, helm) have in common with [the following lines](https://github.com/operator-framework/operator-sdk/blob/master/internal/plugins/helm/v1/scaffolds/internal/templates/makefile.go#:~:text=docker%20push%20%24%7BIMG%7D-,%23%23%40%20Deployment,endif,-.PHONY%3A%20helm%2Doperator):
```shell
.PHONY: install
install: kustomize ## Install CRDs into the K8s cluster specified in ~/.kube/config.
	$(KUSTOMIZE) build config/crd | kubectl apply -f -
...
.PHONY: deploy
deploy: kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
	$(KUSTOMIZE) build config/default | kubectl apply -f -
...
```
So, if I'm understanding correctly, the language specific plugin:
1) assumes there is a `config/` folder
2) then defines `Makefile`
3) and determine how the CRD and its controller get installed.

In general, the `config/` is generated by `kustomize plugin`. (Though the language specific plugin does not care who create that folder.)

Consequently, I have this basic question:
> Once the alpha plugin get triggered by `kb init` and generates:
> - kubebuilderconfiggen.yaml
> - controller_manager_config.yaml
> 
> How can users use these files to apply their configured controllers to the cluster?

Reply:
The temporary approach in the this commit is: 
Add this Makefile-Scaffolding template to the alpha plugin. Then the ultimate `Makefile` will have a difference in its `make deploy`:
```shell
deploy: ## Deploy controller to the K8s cluster specified in ~/.kube/config.
	kubebuilder alpha config-gen ./kubebuilderconfiggen.yaml | kubectl apply -f -
```
That's how `config-gen` get used.

Reply:
This might not be a good one considering on the usage as:
```shell
kubebuilder init --plugins=config-gen/v1-alpha,<language-specific-plugin> ...
```

That makes the alpha plugin an alternative to the kustomize plugin.

IIUIC, the language plugins expect the `config/` folder, regardless of who create that.

So maybe the alpha plugin should generate the `config/`.

Then the language specific plugins won't necessary do/care additional things.

What do you think?

Reply:
Hi @Kavinjsir, 

Thank you for working on this one, Following the comments inline.

> I understand that we don't expect the plugin to "update" makefile, as that is defined as "language specific".
And, it seems that most of the makefile templates(go/v3, go/v2, helm) have in common with [the following lines](https://github.com/operator-framework/operator-sdk/blob/master/internal/plugins/helm/v1/scaffolds/internal/templates/makefile.go#:~:text=docker%20push%20%24%7BIMG%7D-,%23%23%40%20Deployment,endif,-.PHONY%3A%20helm%2Doperator):

.The Makefile is scaffold by the language plugin. 
Why it has specific targets which are by language (e.g controller-gen only work with Go, go fmt, and etc..)

Reply:
> This might not be a good one considering on the usage as:

> kubebuilder init --plugins=config-gen/v1-alpha,<language-specific-plugin> ...
> That makes the alpha plugin an alternative to the kustomize plugin.

It is the goal. The purpose of config-gen would be to replace kustomize
That means users could choose if prefer the config-gen or kustomize approach.